### PR TITLE
feat(eif): add EIF image format support for Nitro Enclaves

### PR DESCRIPTION
**Pull request by jmt-lab**
Introduce a new `image-format = "eif"` variant type that produces an AWS Nitro Enclaves Image Format (kernel + cmdline + ramdisk + metadata) alongside a GPT disk image and bare kernel.

- Add `tools/eif-builder` crate implementing the EIF v4 container format
- Add `twoliter/embedded/rpm2eif` build script and wire it into `build.Dockerfile`, branching on `IMAGE_FORMAT=eif` to skip rpm2img's partitioning/secure-boot flags (rpm2eif owns the layout)
- Extend `buildsys::manifest::ImageFormat` with `Eif` and plumb it through `DockerBuild`; reject `repack-variant` for EIF with a clear error
- Teach `pubsys fetch_variant` to fetch the `.eif`, `-disk.img`, and `-kernel` artifacts published by EIF variants
- Bundle `eif-builder` as a twoliter embedded tool
- Add `hello-eif` test variant and manifest-parsing tests

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

File: twoliter/embedded/rpm2eif [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    shopt -u failglob
+    sbom_rpms=( "${SBOM_PACKAGE_DIR}"/*.rpm )
+    shopt -s failglob
+    if [[ -e "${sbom_rpms[0]}" ]]; then
+        rpm -iv --ignorearch --root "${INSTALL_ROOT}" "${sbom_rpms[@]}"
+    fi
+fi
+
+echo "=== Extracting kernel ==="
+# Deterministic kernel selection: require exactly one vmlinuz.
```
Comment (jmt-lab) [inline twoliter/embedded/rpm2eif]: typo this should be vmlinux, we do not want compressed kernel

File: tools/eif-builder/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+impl std::error::Error for EifError {}
+
+fn eif_arch_flags() -> u16 {
+    if cfg!(target_arch = "aarch64") {
+        EIF_ARCH_AARCH64
+    } else {
+        EIF_ARCH_X86_64
+    }
+}
```
Comment (ginglis13) [inline tools/eif-builder/src/lib.rs]: This is deriving from the build host architecture, not necessarily the target architecture.  let's write through a target arch in `rpm2eif`

File: twoliter/embedded/rpm2eif [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+# Write rootfs and verity data to partitions
+dd if="${ROOTFS_TMP}" of="${DISK_TMP}" conv=notrunc bs=1M seek="${ROOT_OFF_MIB}"
+dd if="${VERITY_TMP}" of="${DISK_TMP}" conv=notrunc bs=1M seek="${HASH_OFF_MIB}"
+
+sgdisk -v "${DISK_TMP}"
+
+echo "=== Building EIF with dm-verity cmdline ==="
+# Build dm-verity kernel cmdline with /dev/vda1 and /dev/vda2
+DM_CREATE="dm-mod.create=\"root,,,ro,0 ${VERITY_DATA_SECTORS} verity 1 /dev/vda1 /dev/vda2 4096 4096 ${VERITY_DATA_BLOCKS} 1 sha256 ${VERITY_ROOT_HASH} ${VERITY_SALT} 2 restart_on_corruption ignore_zero_blocks\""
+KERNEL_CMDLINE="root=/dev/dm-0 ro rootwait console=ttyS0 initcall_blacklist=i8042_init ${DM_CREATE}"
```
Comment (ginglis13) [inline twoliter/embedded/rpm2eif]: should we append `${KERNEL_PARAMETERS}` here incase those are set in the variant definition?

File: twoliter/embedded/rpm2eif [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+# Partition layout: 1MiB GPT gap, ROOT-A, HASH-A
+GPT_GAP_MIB=1
+ROOT_OFF_MIB="${GPT_GAP_MIB}"
+ROOT_SIZE_MIB="${ROOTFS_MIB}"
+HASH_OFF_MIB="$(( ROOT_OFF_MIB + ROOT_SIZE_MIB ))"
+HASH_SIZE_MIB="${VERITY_MIB}"
+DISK_SIZE_MIB="$(( HASH_OFF_MIB + HASH_SIZE_MIB + 1 ))"  # +1 for GPT backup
+
+# Linux discoverable-partitions GUIDs (per the UAPI Group spec) are
+# arch-specific. Fail fast on architectures we have not validated.
```
Comment (ginglis13) [inline twoliter/embedded/rpm2eif]: nit: the x86 values here are the Bottlerocket GUIDs from partyplanner, not the UAPI standard GUIDs

File: twoliter/embedded/rpm2eif [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    exit 1
+fi
+KERNEL="${KERNELS[0]}"
+echo "Kernel: ${KERNEL}"
+
+mapfile -t VMLINUXES < <(find "${INSTALL_ROOT}" -name 'vmlinux' | sort)
+VMLINUX=""
+if (( ${#VMLINUXES[@]} > 0 )); then
+    VMLINUX="${VMLINUXES[0]}"
+    cp "${VMLINUX}" "${OUTPUT_DIR}/vmlinux"
```
Comment (ginglis13) [inline twoliter/embedded/rpm2eif]: should we instead have this as a versioned artifact to avoid collisions across builds?

File: twoliter/embedded/rpm2eif [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+cp "${KERNEL_LOCAL}" "${OUTPUT_DIR}/${ARTIFACT_PREFIX}-kernel"
+cp "${EIF_TMP}" "${OUTPUT_DIR}/${ARTIFACT_PREFIX}.eif"
+cp "${DISK_TMP}" "${OUTPUT_DIR}/${ARTIFACT_PREFIX}-disk.img"
+
+# Create symlinks
+ln -snf "${ARTIFACT_PREFIX}.eif" "${OUTPUT_DIR}/latest.eif"
+ln -snf "${ARTIFACT_PREFIX}-disk.img" "${OUTPUT_DIR}/latest-disk.img"
+ln -snf "${ARTIFACT_PREFIX}-kernel" "${OUTPUT_DIR}/latest-kernel"
+if [[ -n "${VMLINUX}" ]]; then
+    ln -snf "vmlinux" "${OUTPUT_DIR}/latest-vmlinux"
```
Comment (ginglis13) [inline twoliter/embedded/rpm2eif]: should we instead have this as a versioned artifact to avoid collisions across builds?

**Comment by jmt-lab** [comment]
Updated per comments and reference eif file.

File: tools/eif-builder/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+        "ImageVersion": "",
+        "BuildMetadata": {
+            "BuildTime": "",
+            "BuildTool": "twoliter",
+            "BuildToolVersion": build_tool_version,
+            "OperatingSystem": "Linux",
+            "KernelVersion": kernel_version,
+        },
+        "DockerInfo": {},
+        "CustomMetadata": serde_json::Value::Null,
```
Comment (ShelbyZ) [inline tools/eif-builder/src/lib.rs:131]: Should this be `{}` to match `object` - https://github.com/aws/aws-nitro-enclaves-image-format/blob/5704eb0df3986acdb87482e3f37cc869a5c50cf3/src/defs/mod.rs#L281-L282

Comment (jmt-lab) [reply tools/eif-builder/src/lib.rs:131]: It looks like actually null is the correct setting here, it matches nitro-cli's enclave_build which sets custom_info = json!(null) when no --metadata is provided.

Comment (jmt-lab) [reply tools/eif-builder/src/lib.rs:131]: codesnippet from aws-nitro-enclaves-cli enclave_build/src/lib.rs:122
```
let mut custom_info = json!(null);
if let Some(ref path) = self.metadata_path {
    custom_info = parse_custom_metadata(path).map_err(Docker2EifError::MetadataError)?
}
```

File: twoliter/embedded/rpm2eif [hunk 8]
```diff
@@ -0,0 +1,239 @@
+#!/usr/bin/env bash
```
Comment (ginglis13) [inline twoliter/embedded/rpm2eif:1]: Let's generate an application-inventory.json file for EIFs as well

File: twoliter/embedded/partyplanner [hunk 9]
```diff
@@ -413,6 +413,112 @@ set_partition_sizes() {
...
+  # above only rejects empty strings; explicit numeric checks catch
+  # non-numeric input and negative sizes.
+  local n
+  for n in "${rootfs_mib}" "${verity_mib}" "${target_mib}"; do
+    if ! [[ "${n}" =~ ^[0-9]+$ ]]; then
+      echo "set_eif_partition_sizes: expected non-negative integer, got '${n}'" >&2
+      exit 1
+    fi
+  done
+  if (( rootfs_mib == 0 )) || (( verity_mib == 0 )); then
```
Comment (sam-berning) [inline twoliter/embedded/partyplanner]: nit: Might as well check `<= 0` here. 0 *should* be the only possible value with the regex above, but being defensive here may prevent us from merging a bug in the future

File: twoliter/embedded/rpm2eif [hunk 10]
```diff
@@ -0,0 +1,670 @@
...
+    usage
+fi
+
+# Validate the feature flags: EIF has a fixed feature profile, so any flag
+# that doesn't match the EIF-only legal value gets rejected with a clear,
+# variant-author-friendly error. This mirrors the buildsys-side validator in
+# `tools/buildsys/src/manifest.rs::validate_image_features`, which normally
+# catches these earlier -- but rpm2eif can be invoked from other paths (dev
+# scripts, ad-hoc runs), so we defense-in-depth here too.
+declare -A eif_flag_whitelist=(
```
Comment (sam-berning) [inline twoliter/embedded/rpm2eif]: Prefer a name like `eif_flag_allowlist` here.

File: twoliter/embedded/rpm2eif [hunk 11]
```diff
@@ -0,0 +1,670 @@
...
+if [[ -z "${KERNEL_VERSION}" || "${KERNEL_VERSION}" == *"no package"* ]]; then
+    KERNEL_VERSION="unknown"
+fi
+echo "Kernel version: ${KERNEL_VERSION}"
+
+# Copy kernel to local temp before removing /boot from rootfs
+KERNEL_LOCAL="$(mktemp)"
+cp "${KERNEL}" "${KERNEL_LOCAL}"
+
+echo "=== Generating application inventory ==="
```
Comment (sam-berning) [inline twoliter/embedded/rpm2eif:326]: This logic is duplicated from `rpm2img`. Thoughts on lifting this out into a function in `imghelper` so that the logic doesn't drift if it's ever updated?

File: twoliter/embedded/tests/test_rpm2eif_args.sh [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+
+echo
+echo "Test 10: accept --os-image-size-gib=0 (auto-size)"
+assert_passes_validation "os-image-size-gib=0 accepted" \
+  "${BASE[@]}" --os-image-size-gib=0
+
+echo
+echo "Test 11: accept --os-image-size-gib=2"
+assert_passes_validation "os-image-size-gib=2 accepted" \
+  "${BASE[@]}" --os-image-size-gib=2
```
Comment (sam-berning) [inline twoliter/embedded/tests/test_rpm2eif_args.sh:135]: It'd be useful to have more test cases here that are numbers, but we expect to fail. Can you add at a minimum a test for a negative integer and a positive non-integer?

File: tools/buildsys/src/manifest.rs [hunk 13]
```diff
@@ -958,30 +1010,86 @@ pub fn resolved_image_layout(
...
+    let reason_host_containers: String = if is_eif {
+        "host-containers require the BOTTLEROCKET-DATA partition, which an EIF image does not have"
+            .to_string()
+    } else {
+        "host-containers require the BOTTLEROCKET-DATA partition, which is not built when `first-party-stack = false`".to_string()
+    };
+    let reason_first_party_stack =
+        "the EIF pipeline is inherently stripped down; `rpm2eif` requires `first-party-stack = false`";
+    let reason_xfs_data_partition =
+        "xfs-data-partition requires the BOTTLEROCKET-DATA partition, which an EIF image does not have";
```
Comment (sam-berning) [inline tools/buildsys/src/manifest.rs]: Should this also be denied for `first-party-stack` if this is the reason? Both features don't have this partition.

File: tools/eif-builder/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,452 @@
...
+///
+/// The schema (top-level keys and `BuildMetadata` sub-keys) matches
+/// `aws-nitro-enclaves-image-format`. Values are left as empty strings where
+/// we do not populate them; the keys themselves must be present.
+///
+/// NOTE: This intentionally omits the PCR/measurement fields produced by
+/// the upstream AWS Nitro Enclaves CLI. The rootfs is supplied out-of-band
+/// as a virtio-blk device, so attestation of the rootfs cannot be derived
+/// from the EIF alone. Consumers MUST NOT rely on EIF metadata for
+/// attestation of artifacts produced by this builder.
```
Comment (ginglis13) [inline tools/eif-builder/src/lib.rs]: is this true? won't the EIF include the dm-verity hash?

File: tools/eif-builder/patches/.gitkeep
Comment (ginglis13) [inline tools/eif-builder/patches/.gitkeep:1]: what is this for?

File: tools/eif-builder/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,452 @@
...
+        path: PathBuf,
+        source: std::io::Error,
+    },
+}
+
+// NOTE: The EIF header's architecture flag identifies the *target* arch of
+// the kernel embedded in the EifSectionKernel section, not the arch of the
+// machine building the EIF. Cross-arch builds are common (e.g. an aarch64
+// builder producing an x86_64 EIF, or vice-versa), so callers must pass the
+// target arch explicitly via `TargetArch`.
```
Comment (ginglis13) [inline tools/eif-builder/src/lib.rs]: this comment seems useful, but not necessarily in this location; it's dangling here. This is more or less already included in `build_eif`

File: tools/eif-builder/src/main.rs [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+
+#[derive(Parser)]
+#[command(name = "eif-builder", about = "Build a minimal sidecar EIF")]
+struct Args {
+    /// Path to the kernel image to embed. Format is arch-dependent:
+    /// on x86_64 this is an uncompressed ELF vmlinux (Firecracker PVH boot);
+    /// on aarch64 this is either a flat PE-wrapped arm64 kernel Image or an
+    /// EFI zboot image wrapping one (arm64 Linux boot protocol). zboot images
+    /// are unwrapped and decompressed transparently (gzip and zstd supported)
+    /// so the EIF always embeds a bootable flat Image.
```
Comment (ginglis13) [inline tools/eif-builder/src/main.rs]: this comment details and others has been repeated across many places in this PR. Why not have a README under `eif-builder` instead ?

File: tools/eif-builder/patches/.gitkeep
Comment (jmt-lab) [reply tools/eif-builder/patches/.gitkeep:1]: Not sure this is needed anymore i think maybe AI slop generated.

File: tools/eif-builder/src/lib.rs (hunk 14 above)
Comment (jmt-lab) [reply tools/eif-builder/src/lib.rs]: Yeah it does contain the dm-verity hash

**Review by ginglis13 — APPROVED** [review]

File: twoliter/embedded/tests/test_rpm2eif_args.sh [hunk 17]
```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env bash
+#
+# Test the argument-validation half of `rpm2eif`: check that the EIF flag
+# whitelist rejects incompatible feature combinations with clear errors,
```
Comment (sam-berning) [inline twoliter/embedded/tests/test_rpm2eif_args.sh]: nit: Can you update this to allowlist in this file as well?

**Review by sam-berning — APPROVED** [review]
